### ISSUE TITLE
[Perf request] Make index_select on sparse COO tensors as fast as that from rusty1s/pytorch_sparse (1000x)

### ISSUE THREAD

**Issue by tvercaut** [pytorch/pytorch#72212]
### 🚀 The feature, motivation and pitch

The native `index_select` on sparse COO tensors has been reported to be slow in the past (see #61788 and #54561). While improvements have been made with #63008, it remains slower than expected even with pytorch 1.10.0 and 1.10.1 which, as per the [1.10.0 release notes](https://github.com/pytorch/pytorch/releases/tag/v1.10.0), I understand include the #63008 patch.

### Alternatives

The `index_select` implementation from [rusty1s/torch_sparse](https://github.com/rusty1s/pytorch_sparse/blob/master/torch_sparse/index_select.py) can be used in lieu of the native `index_select` and it is sometimes up to 1000x faster.

### Additional context

Below is the simple test case I used as a quick and dirty "benchmark" (also on [colab](https://colab.research.google.com/drive/1-tYbh_KP50NDhBuJ_m_DulRwCAi_ewZO?usp=sharing)):
```python
import torch
print(torch.__version__)

torchdevice = torch.device('cpu')
if torch.cuda.is_available():
  torchdevice = torch.device('cuda')
  print('Default GPU is ' + torch.cuda.get_device_name(torch.device('cuda')))
print('Running on ' + str(torchdevice))

!pip install torch-scatter torch-sparse -f https://data.pyg.org/whl/torch-{torch.__version__}.html
import torch_sparse

# Convenience wrapper around torch_sparse index_select
def ts_index_select(A,sdim,idx):
  Ats = torch_sparse.SparseTensor.from_torch_sparse_coo_tensor(A)
  Ats_select = torch_sparse.index_select(Ats,sdim,idx)
  row, col, value = Ats_select.coo()
  As_select = torch.sparse_coo_tensor(torch.stack([row, col], dim=0), value, (Ats_select.size(0), Ats_select.size(1)))
  return As_select

# Dimension of the square sparse matrix
n = 1000000
# Number of non-zero elements (up to duplicates)
nnz = 100000
# Number of selected indices (up to duplicates)
m = 10000

rowidx = torch.randint(low=0, high=n, size=(nnz,), device=torchdevice)
colidx = torch.randint(low=0, high=n, size=(nnz,), device=torchdevice)
itemidx = torch.vstack((rowidx,colidx))
xvalues = torch.randn(nnz, device=torchdevice)
SparseX = torch.sparse_coo_tensor(itemidx, xvalues, size=(n,n)).coalesce()
print('SparseX:',SparseX)

selectrowidx = torch.unique(torch.randint(low=0, high=n, size=(m,), device=torchdevice), sorted=True)

print('\nRunning index_select from PyTorch')
%timeit SparseXsub1 = SparseX.index_select(0,selectrowidx)

print('\nRunning index_select from torch_sparse')
%timeit SparseXsub2 = ts_index_select(SparseX,0,selectrowidx)
```
with output
```
Running index_select from PyTorch
1 loop, best of 5: 2.15 s per loop

Running index_select from torch_sparse
1000 loops, best of 5: 1.69 ms per loop
```

Despite the suggestion in https://github.com/pytorch/pytorch/issues/61788#issuecomment-908538796 to reopen that issue if performance bottlenecks remained , as I not the OP of #61788, I cannot reopen it myself, hence this new feature request.

cc @nikitaved @pearu @cpuhrsch

**Comment by nikitaved** [comment]
I did have a quick look at the implementation. It appears to be quite serial, and asks for some parallelization.

**Comment by nikitaved** [comment]
@tvercaut , are these times for the CPU or GPU. If for the GPU, could you please append `torch.cuda.synchronize()` to `timeit` like this : `t.index_select(...); torch.cuda.synchronize()`? Could you please show the result for both the CPU and GPU for that matter.

**Comment by nikitaved** [comment]
For the CPU the complexity of the implemented algorithm seems like `nnz x len(index)`, we can do that in `nnz + len(index)`.

**Comment by tvercaut** [comment]
The reported timing were for the GPU but I get roughly the same numbers on CPU for this toy example. I have updated my quick and dirty colab notebook to run both GPU and CPU and also add `torch.cuda.synchronize()` on the `%timeit` line for the GPU:
https://colab.research.google.com/drive/1-tYbh_KP50NDhBuJ_m_DulRwCAi_ewZO?usp=sharing

```
Default GPU is Tesla K80

Running index_select from PyTorch on cuda
1 loop, best of 5: 2.27 s per loop

Running index_select from torch_sparse on cuda
1000 loops, best of 5: 1.79 ms per loop

Running index_select from PyTorch on cpu
1 loop, best of 5: 2.28 s per loop

Running index_select from torch_sparse on cpu
100 loops, best of 5: 5.52 ms per loop
```

**Comment by cpuhrsch** [comment]
This is a reasonable request. We can find someone to work on it, but if it's to be done sooner we're also happy to accept a pull request.

**Comment by nikitaved** [comment]
I can work on this for a change.

**Comment by nikitaved** [comment]
@tvercaut , FYI, the CUDA backend converts tensors to the CPU, runs the algo, then converts back. The CPU part is more or less clear, for CUDA I will stick to the CUDA-CPU-CUDA paradigm for now... I am actively working on the issue.

**Comment by tvercaut** [comment]
@nikitaved, in the actual use case I was looking at, I am interested in indexing with `dim=0` only but it's a bit specific and I assume indexing with `dim=1` (and `dim=2`) as well would serve other use cases.

More specifically, my use case involves a sparse matrix L and I am intested in (conceptually) extracting a sub matrix from L@L.T using the same indices for `dim=0` and `dim=1`. In my current approach, I get Lsub by indexing L for `dim=0` and then create the sparse symmetrix sub matrix I want as Lsub@Lsub.T.
 
The toy example shared here was my attempt at providing a minimal example of the performance issue I was facing.

**Comment by cpuhrsch** [comment]
@nikitaved - on CSC support @pearu was planning to work on it next.

**Comment by jamfly** [comment]
Hi @tvercaut,

Have you been tested on an extremely huge table? When I used the index_select in your approach, I encounter the OOM issue. Plus, it cannot deal with one dimension sparse tensor. I had to make it two-dimensional beforehand.

Or, do you know a better way to get the values from a sparse tensor by a sequence of the index? Original index_select is too slow to use. Something like `tensor[sequence]`  is what I want.
Thank you in advance.

**Comment by nikitaved** [comment]
Master now contains an updated and faster version. Note, however, just like the previous version, CUDA inputs are converted to CPU, then the algorithms operates on CPU and moves the result back to CUDA once it is done.
Some benchmarks are available here: https://github.com/pytorch/pytorch/pull/72710#issue-1132586108

**Comment by cpuhrsch** [comment]
@nikitaved also just merged a native CUDA implementation for this with even more improvements.

@tvercaut - the latest nightly should now contain a much more efficient implement for index_select. Would you mind trying it for your use case again?

**Comment by tvercaut** [comment]
Looks great! I confirm the speed up is massive. Many thanks. I'm looking forward to seeing this in 1.12.

**Comment by cpuhrsch** [comment]
@tvercaut - I'm very happy to hear that! Unfortunately the CUDA performance improvements didn't make the branch cut in time so are to be expected in 1.13, but the CPU performance improvements will be available in 1.12.

**Comment by nikitaved** [comment]
@tvercaut , if your environments permits that, you could install a nightly release that mirrors the current state of master.
